system/datatypes/mailinglist.php

```diff
@@ -412,246 +439,344 @@ private function unsubscribeAction() {
-			$num = count($this->getDoc()->findXPath('//address[@status="' . self::ADDRESS_STATUS_CONFIRMED . '"]'));
+		// create form
+		$form = $this->createMailingForm($request->getPostData());
 
-			$path = $this->language . '/' . $this->pagename . '/' . $mailingId;
-			$commands->append(makeButton(__('send'), 'if(confirm(\'' . __('ml-sure-to-send', array("count" => $num)) . '\'))' . makeAction($path, self::FORM_CMD_MAILING_SEND, '')));
-			$commands->append(makeButton(__('ml-test-send'), 'hypha(\''.$path.'\', \''.self::FORM_CMD_MAILING_TEST_SEND.'\', prompt(\'' . __('email') . '\'), $(this).closest(\'form\'));'));
+		// process form if it was posted
+		$form->validateRequiredField(self::FIELD_NAME_SUBJECT);
+		$form->validateRequiredField(self::FIELD_NAME_MESSAGE);
+		if (!empty($form->errors)) {
+			$cancelPath = str_replace('{id}', $mailingId, self::PATH_MAILS_VIEW_ID);
+			$submitPath = str_replace('{id}', $mailingId, self::PATH_MAILS_EDIT_ID);
+			return $this->mailingFormViewRender($form, $cancelPath, $submitPath);
 		}
 
-		return null;
+		$this->xml->lockAndReload();
+
+		/** @var HyphaDomElement $mailing */
+		$mailing = $this->xml->document()->getElementById($mailingId);
+		$mailing->setAttr(self::FIELD_NAME_SUBJECT, $form->dataFor(self::FIELD_NAME_SUBJECT));
+		$mailing->setHtml(wikify_html($form->dataFor(self::FIELD_NAME_MESSAGE)));
+
+		$this->xml->saveAndUnlock();
+
+		// goto view page with notification
+		notify('success', __('ml-successfully-created'));
+		return ['redirect', $this->constructFullPath($this->pagename . '/' . $mailingId)];
 	}
 
 	/**
-	 * @param array $data
+	 * Creates a HTML form object for the subscribers.
 	 *
+	 * @param array $values
 	 * @return WymHTMLForm
 	 */
-	private function createSubscribeForm(array $data = []) {
-		$email = __('email');
-		$emailFieldName = self::FIELD_NAME_EMAIL;
+	private function createSubscribeForm(array $values = []) {
 		$html = <<<EOF
 			<div class="section" style="padding:5px; margin-bottom:5px; position:relative;">
-				<label for="$emailFieldName">$email</label>: <input type="text" id="$emailFieldName" name="$emailFieldName" placeholder="$email" />
+				<label for="[[emailFieldName]]">[[email]]</label>: <input type="text" id="[[emailFieldName]]" name="[[emailFieldName]]" placeholder="[[email]]" />
 			</div>
 EOF;
-		/** @var HyphaDomElement $form */
-		$form = $this->html->createElement('form');
-		/** @var \DOMWrap\Element $elem */
-		$elem = $form->html($html);
 
-		// buttons
-		/** @var NodeList $field */
-		$field = $elem->find('#' . $emailFieldName);
-		$field->parent()->append($this->makeActionButton(__('subscribe'), null, self::FORM_CMD_LIST_SUBSCRIBE));
+		$vars = [
+			'email' => __('email'),
+			'emailFieldName' => self::FIELD_NAME_EMAIL,
+		];
+
+		$html = hypha_substitute($html, $vars);
 
-		return $this->createForm($elem->children(), $data);
+		return new WymHTMLForm($html, $values);
```

Comment:
This can be a normal `HTMLForm`.

Reply:
true, but I would rather return in all the WymHTMLForm, it's almost no overhead and way more consistent.